internal/cmd/agent/controller/bundledeployment_controller.go

```diff
@@ -0,0 +1,241 @@
+	err := r.Get(ctx, req.NamespacedName, bd)
+	if apierrors.IsNotFound(err) {
+		// This actually deletes the helm releases if a bundledeployment is deleted or orphaned
+		logger.V(1).Info("BundleDeployment deleted, cleaning up helm releases")
+		err := r.Cleanup.CleanupReleases(ctx, key, nil)
+		if err != nil {
+			logger.Error(err, "Failed to clean up missing bundledeployment", "key", key)
+		}
+		return ctrl.Result{}, nil
+	} else if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	if bd.Spec.Paused {
+		logger.V(1).Info("Bundle paused, clearing drift detection")
+		err := r.DriftDetect.Clear(req.String())
+
+		return ctrl.Result{}, err
+	}
+
+	merr := []error{}
+
+	// helm deploy the bundledeployment
+	status, err := r.Deployer.DeployBundle(ctx, bd)
+	if err != nil {
+		logger.V(1).Error(err, "Failed to deploy bundle", "status", status)
+
+		// do not use the returned status, instead set the condition and possibly a timestamp
+		bd.Status = setCondition(bd.Status, err, condition.Cond(fleetv1.BundleDeploymentConditionDeployed))
+
+		merr = append(merr, fmt.Errorf("failed deploying bundle: %w", err))
+	} else {
+		bd.Status = setCondition(status, nil, condition.Cond(fleetv1.BundleDeploymentConditionDeployed))
+	}
+
+	// if we can't retrieve the resources, we don't need to try any of the other operations and requeue now
+	resources, err := r.Deployer.Resources(bd.Name, bd.Status.Release)
+	if err != nil {
+		logger.V(1).Info("Failed to retrieve bundledeployment's resources")
+		// merr = append(merr, fmt.Errorf("failed retrieving resources: %w", err))
+		if err := updateStatus(ctx, logger, r.Client, req.NamespacedName, bd.Status); err != nil {
+			merr = append(merr, fmt.Errorf("failed to update the status: %w", err))
+		}
+		return ctrl.Result{}, errutil.NewAggregate(merr)
+	}
+
+	var result ctrl.Result
+	if monitor.ShouldUpdateStatus(bd) {
+		// update the bundledeployment status and check if we deploy an agent, or if we need to trigger drift correction
+		status, err = r.Monitor.UpdateStatus(ctx, bd, resources)
+		if err != nil {
+			logger.Error(err, "Cannot monitor deployed bundle")
+
+			// if there is an error, do not use the returned status from monitor
+			bd.Status = setCondition(bd.Status, err, condition.Cond(fleetv1.BundleDeploymentConditionMonitored))
+			merr = append(merr, fmt.Errorf("failed updating status: %w", err))
+		} else {
+			// we add to the status from deployer.DeployBundle
+			bd.Status = setCondition(status, nil, condition.Cond(fleetv1.BundleDeploymentConditionMonitored))
+		}
+
+		// Run drift correction
+		if len(status.ModifiedStatus) > 0 && bd.Spec.CorrectDrift.Enabled {
+			err = r.Deployer.RemoveExternalChanges(ctx, bd)
+			if err != nil {
+				merr = append(merr, fmt.Errorf("failed reconciling drift: %w", err))
+			}
+		}
+
+		// TODO is this needed with drift correction?
+		if len(bd.Status.ModifiedStatus) > 0 && monitor.ShouldRedeploy(bd) {
+			result.RequeueAfter = durations.MonitorBundleDelay
```

Comment:
why do we want to requeue here?

Reply:
It's old behavior. I'm not sure we can remove it.

Reply:
https://github.com/rancher/fleet/pull/1977/commits/b7810f0d5b9815e650a4273caa502349bd4fdaf5